File: android/app/src/main/kotlin/com/workoutnotes/workout_notes/run/RunVoiceSettings.kt
package com.workoutnotes.workout_notes.run

import org.json.JSONObject

enum class RunIntervalMetric { distance, time }

data class RunIntervalPreset(
    val workMetric: RunIntervalMetric = RunIntervalMetric.distance,
    val workValue: Int = 400,
    val restMetric: RunIntervalMetric = RunIntervalMetric.time,
    val restValue: Int = 90,
    val repeats: Int = 8,
) {
    companion object {
        fun fromJson(json: JSONObject?): RunIntervalPreset {
            if (json == null) return RunIntervalPreset()
            val workMetric = if (json.optString("workMetric") == "time") RunIntervalMetric.time else RunIntervalMetric.distance
            val restMetric = if (json.optString("restMetric") == "time") RunIntervalMetric.time else RunIntervalMetric.distance
            val workValue = json.optInt("workValue", 400).coerceAtLeast(1)
            val restValue = json.optInt("restValue", 90).coerceAtLeast(0)
            val repeats = json.optInt("repeats", 8).coerceIn(1, 99)
            return RunIntervalPreset(workMetric, workValue, restMetric, restValue, repeats)
        }

        fun fromMap(map: Map<String, Any?>?): RunIntervalPreset {
            if (map == null) return RunIntervalPreset()
            val workMetric = if (map["workMetric"] == "time") RunIntervalMetric.time else RunIntervalMetric.distance
            val restMetric = if (map["restMetric"] == "time") RunIntervalMetric.time else RunIntervalMetric.distance
            val workValue = (map["workValue"] as? Number)?.toInt()?.coerceAtLeast(1) ?: 400
            val restValue = (map["restValue"] as? Number)?.toInt()?.coerceAtLeast(0) ?: 90
            val repeats = (map["repeats"] as? Number)?.toInt()?.coerceIn(1, 99) ?: 8
            return RunIntervalPreset(workMetric, workValue, restMetric, restValue, repeats)
        }
    }

    fun toJson(): JSONObject = JSONObject().apply {
        put("workMetric", workMetric.name)
        put("workValue", workValue)
        put("restMetric", restMetric.name)
        put("restValue", restValue)
        put("repeats", repeats)
    }
}

data class RunVoiceSettings(
    val enabled: Boolean = true,
    val headphonesOnly: Boolean = true,
    val muteDuringCall: Boolean = true,
    val announceDistance: Boolean = true,
    val distanceEveryKm: Int = 1,
    val announceSplit: Boolean = true,
    val announcePaceWarning: Boolean = false,
    val targetPaceSecPerKm: Int? = null,
    val paceTolerancePercent: Int = 10,
    val announceGpsStatus: Boolean = true,
    val announceIntervals: Boolean = true,
    val intervalsEnabledByDefault: Boolean = false,
    val interval: RunIntervalPreset = RunIntervalPreset(),
) {
    companion object {
        const val STORAGE_KEY = "run_voice_settings_v1"

        fun defaults() = RunVoiceSettings()

        fun fromJsonString(raw: String?): RunVoiceSettings {
            if (raw.isNullOrBlank()) return defaults()
            return try {
                val json = JSONObject(raw)
                fromJson(json)
            } catch (_: Throwable) {
                defaults()
            }
        }

        fun fromJson(json: JSONObject?): RunVoiceSettings {
            if (json == null) return defaults()
            val everyRaw = json.optInt("distanceEveryKm", 1)
            val every = if (everyRaw == 2 || everyRaw == 5) everyRaw else 1
            val tolerance = json.optInt("paceTolerancePercent", 10).coerceIn(5, 50)
            val target = if (json.isNull("targetPaceSecPerKm")) null else json.optInt("targetPaceSecPerKm").takeIf { it > 0 }
            val intervalObj = if (json.has("interval") && !json.isNull("interval")) {
                try { json.getJSONObject("interval") } catch (_: Throwable) { null }
            } else null
            return RunVoiceSettings(
                enabled = json.optBoolean("enabled", true),
                headphonesOnly = json.optBoolean("headphonesOnly", true),
                muteDuringCall = json.optBoolean("muteDuringCall", true),
                announceDistance = json.optBoolean("announceDistance", true),
                distanceEveryKm = every,
                announceSplit = json.optBoolean("announceSplit", true),
                announcePaceWarning = json.optBoolean("announcePaceWarning", false),
                targetPaceSecPerKm = target,
                paceTolerancePercent = tolerance,
                announceGpsStatus = json.optBoolean("announceGpsStatus", true),
                announceIntervals = json.optBoolean("announceIntervals", true),
                intervalsEnabledByDefault = json.optBoolean("intervalsEnabledByDefault", false),
                interval = RunIntervalPreset.fromJson(intervalObj),
            )
        }

        fun fromMap(map: Map<String, Any?>?): RunVoiceSettings {
            if (map == null) return defaults()
            val everyRaw = (map["distanceEveryKm"] as? Number)?.toInt() ?: 1
            val every = if (everyRaw == 2 || everyRaw == 5) everyRaw else 1
            val tolerance = (map["paceTolerancePercent"] as? Number)?.toInt()?.coerceIn(5, 50) ?: 10
            val target = (map["targetPaceSecPerKm"] as? Number)?.toInt()?.takeIf { it > 0 }
            @Suppress("UNCHECKED_CAST")
            val intervalRaw = map["interval"] as? Map<String, Any?>
            return RunVoiceSettings(
                enabled = map["enabled"] as? Boolean ?: true,
                headphonesOnly = map["headphonesOnly"] as? Boolean ?: true,
                muteDuringCall = map["muteDuringCall"] as? Boolean ?: true,
                announceDistance = map["announceDistance"] as? Boolean ?: true,
                distanceEveryKm = every,
                announceSplit = map["announceSplit"] as? Boolean ?: true,
                announcePaceWarning = map["announcePaceWarning"] as? Boolean ?: false,
                targetPaceSecPerKm = target,
                paceTolerancePercent = tolerance,
                announceGpsStatus = map["announceGpsStatus"] as? Boolean ?: true,
                announceIntervals = map["announceIntervals"] as? Boolean ?: true,
                intervalsEnabledByDefault = map["intervalsEnabledByDefault"] as? Boolean ?: false,
                interval = RunIntervalPreset.fromMap(intervalRaw),
            )
        }
    }

    fun toJson(): JSONObject = JSONObject().apply {
        put("enabled", enabled)
        put("headphonesOnly", headphonesOnly)
        put("muteDuringCall", muteDuringCall)
        put("announceDistance", announceDistance)
        put("distanceEveryKm", distanceEveryKm)
        put("announceSplit", announceSplit)
        put("announcePaceWarning", announcePaceWarning)
        if (targetPaceSecPerKm != null) put("targetPaceSecPerKm", targetPaceSecPerKm) else put("targetPaceSecPerKm", JSONObject.NULL)
        put("paceTolerancePercent", paceTolerancePercent)
        put("announceGpsStatus", announceGpsStatus)
        put("announceIntervals", announceIntervals)
        put("intervalsEnabledByDefault", intervalsEnabledByDefault)
        put("interval", interval.toJson())
    }
}

data class RunSessionGoal(
    val enabled: Boolean = false,
    val metric: RunIntervalMetric = RunIntervalMetric.distance,
    val value: Int = 5000,
) {
    companion object {
        fun disabled() = RunSessionGoal(enabled = false)
        fun fromMap(map: Map<String, Any?>?): RunSessionGoal {
            if (map == null) return disabled()
            val enabled = map["enabled"] as? Boolean ?: false
            val metric = if (map["metric"] == "time") RunIntervalMetric.time else RunIntervalMetric.distance
            val value = (map["value"] as? Number)?.toInt() ?: 5000
            return RunSessionGoal(enabled, metric, value)
        }

        fun fromJson(json: JSONObject?): RunSessionGoal {
            if (json == null) return disabled()
            val enabled = json.optBoolean("enabled", false)
            val metric = if (json.optString("metric") == "time") RunIntervalMetric.time else RunIntervalMetric.distance
            val value = json.optInt("value", 5000)
            return RunSessionGoal(enabled, metric, value)
        }
    }

    fun isComplete(distanceMeters: Double, movingTimeSeconds: Int): Boolean {
        if (!enabled || value <= 0) return false
        return if (metric == RunIntervalMetric.distance) {
            distanceMeters + 1e-6 >= value
        } else {
            movingTimeSeconds >= value
        }
    }

    fun toJson(): JSONObject = JSONObject().apply {
        put("enabled", enabled)
        put("metric", metric.name)
        put("value", value)
    }
}

File: lib/models/run_voice_settings.dart
enum RunIntervalMetric { distance, time }

class RunIntervalPreset {
  final RunIntervalMetric workMetric;
  final int workValue;
  final RunIntervalMetric restMetric;
  final int restValue;
  final int repeats;

  const RunIntervalPreset({
    required this.workMetric,
    required this.workValue,
    required this.restMetric,
    required this.restValue,
    required this.repeats,
  });

  /// Default: 400 m work / 90 s rest × 8.
  const RunIntervalPreset.defaults()
      : this(
          workMetric: RunIntervalMetric.distance,
          workValue: 400,
          restMetric: RunIntervalMetric.time,
          restValue: 90,
          repeats: 8,
        );

  RunIntervalPreset copyWith({
    RunIntervalMetric? workMetric,
    int? workValue,
    RunIntervalMetric? restMetric,
    int? restValue,
    int? repeats,
  }) {
    return RunIntervalPreset(
      workMetric: workMetric ?? this.workMetric,
      workValue: workValue ?? this.workValue,
      restMetric: restMetric ?? this.restMetric,
      restValue: restValue ?? this.restValue,
      repeats: repeats ?? this.repeats,
    );
  }

  Map<String, dynamic> toJson() => {
        'workMetric': workMetric.name,
        'workValue': workValue,
        'restMetric': restMetric.name,
        'restValue': restValue,
        'repeats': repeats,
      };

  factory RunIntervalPreset.fromJson(Map<String, dynamic>? json) {
    if (json == null) return const RunIntervalPreset.defaults();
    return RunIntervalPreset(
      workMetric: _metric(json['workMetric']),
      workValue: _positiveInt(json['workValue'], 400),
      restMetric: _metric(json['restMetric']),
      restValue: _nonNegativeInt(json['restValue'], 90),
      repeats: _positiveInt(json['repeats'], 8).clamp(1, 99),
    );
  }

  static RunIntervalMetric _metric(Object? raw) {
    if (raw == 'time') return RunIntervalMetric.time;
    return RunIntervalMetric.distance;
  }

  static int _positiveInt(Object? raw, int fallback) {
    final n = (raw as num?)?.toInt();
    if (n == null || n <= 0) return fallback;
    return n;
  }

  static int _nonNegativeInt(Object? raw, int fallback) {
    final n = (raw as num?)?.toInt();
    if (n == null || n < 0) return fallback;
    return n;
  }
}

class RunVoiceSettings {
  static const storageKey = 'run_voice_settings_v1';

  final bool enabled;
  final bool headphonesOnly;
  final bool muteDuringCall;
  final bool announceDistance;
  final int distanceEveryKm;
  final bool announceSplit;
  final bool announcePaceWarning;
  final int? targetPaceSecPerKm;
  final int paceTolerancePercent;
  final bool announceGpsStatus;
  final bool announceIntervals;
  final bool intervalsEnabledByDefault;
  final RunIntervalPreset interval;

  const RunVoiceSettings({
    required this.enabled,
    required this.headphonesOnly,
    required this.muteDuringCall,
    required this.announceDistance,
    required this.distanceEveryKm,
    required this.announceSplit,
    required this.announcePaceWarning,
    required this.targetPaceSecPerKm,
    required this.paceTolerancePercent,
    required this.announceGpsStatus,
    required this.announceIntervals,
    required this.intervalsEnabledByDefault,
    required this.interval,
  });

  const RunVoiceSettings.defaults()
      : this(
          enabled: true,
          headphonesOnly: true,
          muteDuringCall: true,
          announceDistance: true,
          distanceEveryKm: 1,
          announceSplit: true,
          announcePaceWarning: false,
          targetPaceSecPerKm: null,
          paceTolerancePercent: 10,
          announceGpsStatus: true,
          announceIntervals: true,
          intervalsEnabledByDefault: false,
          interval: const RunIntervalPreset.defaults(),
        );

  RunVoiceSettings copyWith({
    bool? enabled,
    bool? headphonesOnly,
    bool? muteDuringCall,
    bool? announceDistance,
    int? distanceEveryKm,
    bool? announceSplit,
    bool? announcePaceWarning,
    int? targetPaceSecPerKm,
    bool clearTargetPace = false,
    int? paceTolerancePercent,
    bool? announceGpsStatus,
    bool? announceIntervals,
    bool? intervalsEnabledByDefault,
    RunIntervalPreset? interval,
  }) {
    return RunVoiceSettings(
      enabled: enabled ?? this.enabled,
      headphonesOnly: headphonesOnly ?? this.headphonesOnly,
      muteDuringCall: muteDuringCall ?? this.muteDuringCall,
      announceDistance: announceDistance ?? this.announceDistance,
      distanceEveryKm: distanceEveryKm ?? this.distanceEveryKm,
      announceSplit: announceSplit ?? this.announceSplit,
      announcePaceWarning: announcePaceWarning ?? this.announcePaceWarning,
      targetPaceSecPerKm: clearTargetPace
          ? null
          : (targetPaceSecPerKm ?? this.targetPaceSecPerKm),
      paceTolerancePercent: paceTolerancePercent ?? this.paceTolerancePercent,
      announceGpsStatus: announceGpsStatus ?? this.announceGpsStatus,
      announceIntervals: announceIntervals ?? this.announceIntervals,
      intervalsEnabledByDefault:
          intervalsEnabledByDefault ?? this.intervalsEnabledByDefault,
      interval: interval ?? this.interval,
    );
  }

  Map<String, dynamic> toJson() => {
        'enabled': enabled,
        'headphonesOnly': headphonesOnly,
        'muteDuringCall': muteDuringCall,
        'announceDistance': announceDistance,
        'distanceEveryKm': distanceEveryKm,
        'announceSplit': announceSplit,
        'announcePaceWarning': announcePaceWarning,
        'targetPaceSecPerKm': targetPaceSecPerKm,
        'paceTolerancePercent': paceTolerancePercent,
        'announceGpsStatus': announceGpsStatus,
        'announceIntervals': announceIntervals,
        'intervalsEnabledByDefault': intervalsEnabledByDefault,
        'interval': interval.toJson(),
      };

  factory RunVoiceSettings.fromJson(Map<String, dynamic>? json) {
    if (json == null) return const RunVoiceSettings.defaults();
    final every = (json['distanceEveryKm'] as num?)?.toInt() ?? 1;
    final tolerance = (json['paceTolerancePercent'] as num?)?.toInt() ?? 10;
    final target = (json['targetPaceSecPerKm'] as num?)?.toInt();
    final intervalRaw = json['interval'];
    return RunVoiceSettings(
      enabled: json['enabled'] as bool? ?? true,
      headphonesOnly: json['headphonesOnly'] as bool? ?? true,
      muteDuringCall: json['muteDuringCall'] as bool? ?? true,
      announceDistance: json['announceDistance'] as bool? ?? true,
      distanceEveryKm: every == 2 || every == 5 ? every : 1,
      announceSplit: json['announceSplit'] as bool? ?? true,
      announcePaceWarning: json['announcePaceWarning'] as bool? ?? false,
      targetPaceSecPerKm: target != null && target > 0 ? target : null,
      paceTolerancePercent: tolerance.clamp(5, 50),
      announceGpsStatus: json['announceGpsStatus'] as bool? ?? true,
      announceIntervals: json['announceIntervals'] as bool? ?? true,
      intervalsEnabledByDefault:
          json['intervalsEnabledByDefault'] as bool? ?? false,
      interval: RunIntervalPreset.fromJson(
        intervalRaw is Map
            ? Map<String, dynamic>.from(intervalRaw)
            : null,
      ),
    );
  }
}
